testsuite/tests/typing-modular-explicits/general.ml

```diff
@@ -0,0 +1,1031 @@
+|}]
+
+(* We cannot omit the module argument like other labelled arguments because of
+possible type dependancy *)
+let apply_labelled_fail = labelled ~y:3
+
+[%%expect{|
+Line 1, characters 26-34:
+1 | let apply_labelled_fail = labelled ~y:3
+                              ^^^^^^^^
+Error: This expression has type "(module M : Typ) -> y:M.t -> M.t"
+       Received an expression argument. However, module arguments cannot be omitted.
+|}]
+
+(* Here we can remove the dependancy, thus it should be accepted. *)
+let labelled' (module M : Typ with type t = int) ~(y:M.t) = y
+
+let apply_labelled_success = labelled' ~y:3
+
+[%%expect{|
+val labelled' : (module M : Typ with type t = int) -> y:M.t -> M.t = <fun>
+val apply_labelled_success : (module Typ with type t = int) -> int = <fun>
+|}]
+
+(* Check that the optionnal argument is removed correclty when applying a
+module argument. *)
+let apply_opt (f : ?opt:int -> (module M : Typ) -> M.t) = f (module Int)
+
+[%%expect{|
+val apply_opt : (?opt:int -> (module M : Typ) -> M.t) -> Int.t = <fun>
+|}]
+
+let build_pair (module M : Typ) ~x ~y : M.t * M.t = (x, y)
+
+let principality_warning = build_pair (module Int) ~y:3 ~x:1
+
+[%%expect{|
+val build_pair : (module M : Typ) -> x:M.t -> y:M.t -> M.t * M.t = <fun>
+val principality_warning : Int.t * Int.t = (1, 3)
+|}]
+
+let foo f a =
+  let _ = (f ~a : (module M : Typ) -> M.t) in
+  f ~a (fun x -> x)
+
+[%%expect{|
+Line 3, characters 7-19:
+3 |   f ~a (fun x -> x)
+           ^^^^^^^^^^^^
+Error: This expression should not be a function, the expected type is
+       "(module Typ)"
+|}]
+
+let foo2 f a =
+  let m = (module Int : Typ) in
+  let _ = (f ~a : (module M : Typ) -> M.t) in
+  f ~a m
+
+[%%expect{|
+Line 4, characters 2-6:
+4 |   f ~a m
+      ^^^^
+Error: This expression has type "(module M : Typ) -> M.t"
+       but an expression was expected of type "(module Typ) -> 'a"
+       The module "M" would escape its scope
+  Attempted to remove dependency because
+  could not extract path from module argument.
+|}]
+
+let foo3 f a =
+  let _ = (f ~a : (module M : Typ) -> M.t) in
+  f ~a (module Int)
+
+[%%expect{|
+Line 3, characters 2-6:
+3 |   f ~a (module Int)
+      ^^^^
+Error: This expression has type "(module M : Typ) -> M.t"
+       but an expression was expected of type "(module Typ) -> 'a"
+       The module "M" would escape its scope
+  Attempted to remove dependency because
+  could not extract path from module argument.
+|}]
+
+(** From here we test possible expressions for the module argument. *)
+
+let x_from_struct = id (module struct type t = int end) 3
+
+[%%expect{|
+val x_from_struct : int = 3
+|}]
+
+module F () : Typ = struct type t = int end
+
+let x_from_generative_functor = id (module F ())
+
+[%%expect{|
+module F : () -> Typ
+Line 3, characters 35-48:
+3 | let x_from_generative_functor = id (module F ())
+                                       ^^^^^^^^^^^^^
+Error:
+       The module "T" would escape its scope
```

Comment:
Do you understand where `"T"` comes from in this error message?

Reply:
The `"T"` comes from the type of id (it is of type `(module T : Typ) -> T.t`)

Reply:
After a bit of looking around it seems that we are missing some information in the trace raised when failing to inline the type of `F ()`